bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/CreateAdvBuilder.java

```diff
@@ -0,0 +1,36 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership. The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *pa
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.bookkeeper.client.api;
+
+import java.util.concurrent.CompletableFuture;
+import org.apache.bookkeeper.client.AsyncCallback.CreateCallback;
+import org.apache.bookkeeper.client.BKException;
+
+/**
+ * Builder-style interface to create new ledgers
+ */
+public interface CreateAdvBuilder {
+
+    CreateAdvBuilder withLedgerId(int ledgerId);
+
+    WriteAdvHandler create() throws BKException, InterruptedException;
+
+    CompletableFuture<WriteAdvHandler> execute();
+
+    void create(CreateCallback cb, Object ctx);
```

Comment:
Why include the old style callbacks when the same can be achieved with the CompletableFuture? Similarly for the synchronous version, you could just do execute().get(). It's better to have fewer paths to maintain, and to encourage people to use the async version.

Reply:
This is to ease the porting of existing applications.
Maybe we could drop the sync version, but IMO most of BK based apps implement the callbacks and switching a whole project.

Additionally using CompletableFuture.get() forces you to handle ExecutionException which is not very friendly

Reply:
The old callback can be added with an adapter, which isn't part of the new interface. We can even provide the adapter in a compat package.

```
createLedgerBuilder().withEnsemble(3).withAckQuorum(2).execute().whenCompleteAsync(CbAdapters.adaptCreate(cb, ctx));
```

Reply:
@ivankelly make sense to me

@sijie what do you think about this ?

Reply:
@ivankelly After thinking about this a bit more I think that the sync method is to be provided.
Dealing with ExecutionException is not very "friendly", it is better to have a way to catch BKException for this kind of operations (create/delete/open).

So for me the best is to have the CompletableFuture version and the sync version, dropping the dependency to the legacy callback.

for addEntry I think we have to keep the callback approach as it is the most GC-friendly async API. I don't know the real footprint of CompletableFuture but it is not "free".
In my opinion we have to decide now the global approach to addEntry as we are designing the new WriteHandler interface and we will maintain this new API in the short/mid term

@merlimat @sijie

Reply:
I'm fine with there being a sync version. Presumably it would just wrap the get and handle the ExecutionException. I think it's more important not to give a new lease of life to the legacy callbacks.

Reply:
@ivankelly  done (dropped legacy async methods)

Reply:
I think fewer methods to execute a operation will be clean and better. so if we can have only one async method, that would be great!

regarding ExecutionException, you can always have a FutureUtils to wait CompletableFuture result and handle ExecutionException. this is how we did in distributedlog.

https://github.com/apache/distributedlog/blob/master/distributedlog-common/src/main/java/org/apache/distributedlog/common/concurrent/FutureUtils.java#L60

Reply:
Very interesting @sijie. I agree and FutureUtils is great.
I will add a FutureUtils in BK and then link to it in javadocs.
This new class will be moved to the commons module which is going to introduce @zhaijai

Reply:
@eolivelli feel free to copy the FutureUtils.java to common/concurrent module.